WX: Comprehensive HiDPI Patch

EmptyChaos opened the pull request:
This is a comprehensive patch for DolphinWX that fixes every HiDPI problem I could find, and also fixes a few tangential segfaults and memory leaks I found along the way. It touches everything in DolphinWX so it's pretty big.
![mainwindow150](https://cloud.githubusercontent.com/assets/17393426/17166720/0539891c-541c-11e6-8eb1-245bea0939d1.png)
[Blue circle: high-res asset. Red circle: Upscaled low-res asset (no high-res available)]

![inputconfigdiag150](https://cloud.githubusercontent.com/assets/17393426/17166728/0f111fb8-541c-11e6-9747-b1832f5e559f.png)

- [x] CFrame / CGameListCtrl
- [x] About Dolphin
- [x] ISOProperties / ARCodeEdit / PatchEdit
- [x] Memory Card Manager
- [x] FifoPlayer Dialog
- [x] Cheat Manager
- [x] Controller Config Dialog
- [x] Input Config Dialog
- [x] General Configuration Dialog
- [x] TAS Input
- [x] Video Configuration Dialog / Postprocess Shader Configuration Dialog
- [x] Debugger Tools
- [x] NetPlay
- [x] Clean build on all platforms

This does *not* include my other PRs (#4048, #4015). It has been tested on Windows and Linux but not OS X, it should work on OS X though.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/4068)
<!-- Reviewable:end -->

JMC47 commented:
Well this is a pleasant surprise.  I don't have a HiDPI monitor to really mess with this on though.

MaJoRoesch commented:
Oh, Wii banners are twice the resolution of GameCube banners, and that could be HiDPI. Looking at the image it doesn't seem like you got that... You do want to try it?

EmptyChaos commented:
@MaJoRoesch I didn't actually know that, but fortunately I just checked and the way I wrote the code is adaptive so it already takes advantage of that fact so I don't need to change anything.

@JMC47 You can manually increase the DPI of your screen in the resolution settings on Windows if you want to mess around with it.

lioncash commented:
Review status: 0 of 98 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/Core/DolphinWX/DolphinSlider.h, line 7 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtmWHgM4ql5cL9Hj18:-KNtmWHhd9Y6g_VeAMiE:784174318) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/DolphinSlider.h#L7)):*
> ```objc
> #pragma once
> 
> #include "wx/slider.h"
> ```

```cpp
#include <wx/slider.h>
```

---

*[Source/Core/DolphinWX/Frame.cpp, line 479 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtmpw-YTye8dYT2YMB:-KNtmpw-YTye8dYT2YMC:1470623051) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/Frame.cpp#L479)):*
> ```cpp
>   // Start debugging maximized (Must be after the window has been positioned)
>   if (UseDebugger)
>     this->Maximize(true);
> ```

`this` isn't necessary (as far as I can tell)

---

*[Source/Core/DolphinWX/Frame.h, line 122 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtnWf8jaCDN_vJS4E7:-KNtnWf9IAzeuv9xwSdH:-1634378989) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/Frame.h#L122)):*
> ```objc
>   const CGameListCtrl* GetGameListCtrl() const;
>   wxMenuBar* GetMenuBar() const override;
>   const wxSize& GetToolbarBitmapSize() const  // Needed before the toolbar exists
> ```

The definition should be in the cpp file for consistency.

---

*[Source/Core/DolphinWX/FrameAui.cpp, line 975 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtnjLWCpIZmyJQCL0d:-KNtnjLXA0JwRV7auirQ:-747530816) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/FrameAui.cpp#L975)):*
> ```cpp
>                                          wxNO_BORDER;
> 
>   wxAuiNotebook* const nb =
> ```

You can just use `auto*` here, since it's obvious what the type is based on the right-hand side of the assignment.

---

*[Source/Core/DolphinWX/Main.h, line 23 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtotJu7LcWkh4P5Oah:-KNtotJu7LcWkh4P5Oai:-1634378989) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/Main.h#L23)):*
> ```objc
> 
>   // Can be called from any thread
>   bool IsActiveThreadsafe() const { return m_is_active_threadsafe.load(std::memory_order_relaxed); }
> ```

The definition should be in the cpp file for consistency.

---

*[Source/Core/DolphinWX/Cheats/CheatSearchTab.cpp, line 103 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtqsZrV2N2BiWod5Mh:-KNtqsZstSgy2Pgd34D8:-1900824458) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/Cheats/CheatSearchTab.cpp#L103)):*
> ```cpp
>   // TODO: Implement between search
>   wxArrayString filters;
>   filters.Add(_("No Filter"));
> ```

The reason this was 'Unknown' was to signify an unknown initial value. Which is the same terminology Cheat Engine uses.

---

*[Source/Core/DolphinWX/Cheats/CheatSearchTab.cpp, line 303 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtqHIFGDGmd3TpA52U:-KNtqHIG99BCExtS2OqO:530347375) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/Cheats/CheatSearchTab.cpp#L303)):*
> ```cpp
>   m_lview_search_results->SetItem(index, 2, buf);
> 
>   // FIXME: Is this actually useful? doubles are 64bits wide.
> ```

This isn't related to DPI changes; neither is the constexpr-ifying below. Please include these in a different PR if necessary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068)*
<!-- Sent from Reviewable.io -->

EmptyChaos commented:
Review status: 0 of 98 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/Core/DolphinWX/DolphinSlider.h, line 7 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtmWHgM4ql5cL9Hj18:-KNtwBzRwFz0QWLFxJnP:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/DolphinSlider.h#L7)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```cpp
> #include <wx/slider.h>
> ```
</details>
Done.

---

*[Source/Core/DolphinWX/Frame.cpp, line 479 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtmpw-YTye8dYT2YMB:-KNtwE50TSlBtN0DnSpz:-1960498722) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/Frame.cpp#L479)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `this` isn't necessary (as far as I can tell)
</details>
The code was cut and pasted from the top of the function. Fixed.

---

*[Source/Core/DolphinWX/Frame.h, line 122 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtnWf8jaCDN_vJS4E7:-KNtwN6VWDaVbJvc9Y7d:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/Frame.h#L122)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> The definition should be in the cpp file for consistency.
</details>
Done.

---

*[Source/Core/DolphinWX/FrameAui.cpp, line 975 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtnjLWCpIZmyJQCL0d:-KNtwProW9XRGlm60Uvw:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/FrameAui.cpp#L975)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You can just use `auto*` here, since it's obvious what the type is based on the right-hand side of the assignment.
</details>
Done.

---

*[Source/Core/DolphinWX/Main.h, line 23 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtotJu7LcWkh4P5Oah:-KNtwSXjw5kYxgSKgHkO:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/Main.h#L23)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> The definition should be in the cpp file for consistency.
</details>
Done.

---

*[Source/Core/DolphinWX/Cheats/CheatSearchTab.cpp, line 103 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtqsZrV2N2BiWod5Mh:-KNtwUbBHjFNVFK3H3yV:-703798473) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/Cheats/CheatSearchTab.cpp#L103)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> The reason this was 'Unknown' was to signify an unknown initial value. Which is the same terminology Cheat Engine uses.
</details>
I'm not sure this is good user design. I had no idea what that did until I read the code, and after reading the code I still have no idea why this option even exists. Reverted.

---

*[Source/Core/DolphinWX/Cheats/CheatSearchTab.cpp, line 303 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068#-KNtqHIFGDGmd3TpA52U:-KNtw3j2zA-6LizsyEdO:1890722128) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6e037454a209772ca373378ad3cfeacef9f79d06/Source/Core/DolphinWX/Cheats/CheatSearchTab.cpp#L303)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This isn't related to DPI changes; neither is the constexpr-ifying below. Please include these in a different PR if necessary.
</details>
They were just incidental things I noticed while working on the surrounding code. Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4068)*
<!-- Sent from Reviewable.io -->

Pringo commented:
@EmptyChaos @MaJoRoesch This might be a good time to add in 2x star images instead of continuing to deal with the low res assets. Do you know if those are Dolphin originals or from some icon pack?

MaJoRoesch commented:
I have noooo idea where they came from. It's probably best to replace them.

I had some difficulty replacing them before, from the small size and complex shape, but I can try again!

Pringo commented:
@MaJoRoesch I tried doing a reverse image search but I haven't been able to find a source for the stars. If the size is a problem we could probably make them taller although they would also need to take up more horizontal space as well if anyone cares about that.

EmptyChaos commented:
@Pringo The images are scaled up using Bicubic filtering (from wxWidgets). At 1x images are not scaled at all, the screenshot is taken at 150% DPI (144DPI). The available filters are bicubic, bilinear, box and point.

If you're using a low-DPI screen to look at the screenshot then it will seem worse than it is because your pixel density is too low, it's important to remember that the physical size of the image in inches/centimeters is the same at high-DPI as the current GUI is in inches/centimeters. i,e. the current images are 1 inch by 1/3 inch, that will remain constant at higher DPI settings.

----

Regarding the image assets, the code I wrote supports quarter fractionals so you can have `xyz@1.5.png`, `xyz@1.25.png` etc. The required image sizes for Windows are `@1`, `@1.25`, `@1.5` and `@2`. I believe OS X requires `@1`, `@2` and `@3`.

Pringo commented:
@EmptyChaos Okay that might be my issue then. I only have a 1080p monitor.

mbc07 commented:
Tested on my laptop (15.5" @ 1080p, Windows scale settings at 125% -- 120 DPI) and while most of the overlapping buttons/controls were fixed, Dolphin still loads low-res bitmaps, making them blurry (except for Wii game banners):
![capturar](https://cloud.githubusercontent.com/assets/8005966/17272150/0ac8654a-565d-11e6-9505-d25925161348.PNG)

Pringo commented:
So Windows will only use the 2x assets if scaling is set to 200% or higher? That is a bit unfortunate. Is there no option to load the 2x asset and downscale for settings between 100% and 200%?

EmptyChaos commented:
Windows doesn't provide any specific support, I had to write all the relevant code so I can change it however is needed. The implementation I went with is to look for an exact match then search for a nearest match so if the DPI is 125% it will look for `@1.25`, `@1.5` then `@1` and scale the nearest match to the correct size. Currently Dolphin only has assets for `@2` and `@1` so for 100% and 125% it will use `@1`, for 150% and 200% it will use `@2`. [This is why I mentioned Windows will need `@1.5` (and, ideally, `@1.25`) resources added]

The theory was that scaling from nearest would produce better results than always scaling down from a bigger size but I could revise the algorithm to more aggressively search for larger sizes so that it will always find the `@2` resources. The results are still likely to be significantly worse than prepared assets though.

Pringo commented:
I think it might be worth trying to load the ceiling of the scaling factor when available, so 1.25x, 1.5x -> 2x. 1x or less -> 1x, etc. Do you know how OS X works with non-integer scaling? I was under the impression that they don't use assets like 1.25x and it gets good results. I generally figured it just downscaled from the next highest version available.

EmptyChaos commented:
I've altered the image search to scan upwards then fallback to smaller sizes if it fails. I've also implemented scale2x/EPX image scaling and applied it to the game banners.
![mainwindow125](https://cloud.githubusercontent.com/assets/17393426/17274267/eca5f5c6-571c-11e6-8a52-9f9367933a35.png)
The results are less blurry than before but there is artifacting if you look closely.

Pringo commented:
@EmptyChaos I'm not sure I'm a fan of the results with scale2x. The banners looked blurry with bicubic but now they look a bit off although somewhat sharper. I'm not sure if this will be any better but would you be willing to try bilinear and see what the results look like with that? I think sharper and maybe somewhat pixely might be fine compromise.

EmptyChaos commented:
@Pringo Here's a full set of combinations:
<img width="300" alt="algorithms" src="https://cloud.githubusercontent.com/assets/17393426/17274439/74ae1438-5724-11e6-8eda-35b5154375a9.png">

Pringo commented:
@EmptyChaos Thanks! Looking at them side by side I think I do like the bilinear results the best. Bicubic is too blurry and the Scale2x variants produce noticeable artifacts in some banners.

EmptyChaos commented:
I've switched the banner scaling to bilinear.
![mainwindow150](https://cloud.githubusercontent.com/assets/17393426/17274639/3fa79142-572e-11e6-8fac-f6ec00341940.png)

Pringo commented:
LGTM

mbc07 commented:
Retested and no more blurry bitmaps for the ones that have higher res available, LGTM too...
![capturar](https://cloud.githubusercontent.com/assets/8005966/17275205/84ddccf6-56cd-11e6-878e-4a225127f70b.PNG)

MaJoRoesch commented:
Did you implement the @2x nobanner? I don't see it anywhere in the files changed tab.

EmptyChaos commented:
@MaJoRoesch The nobanner resource was in an awkward place (ISOFile doesn't have a screen context) so I'd skipped over it. I've fixed the oversight so that it uses the same code path as everything else, Dolphin will locate the `@2` resource for it now.

GhettoGirl commented:
This patches makes everything smaller on my screen (low-res screen: 96dpi 1920x1080). This causes problems on low-dpi screens, while fixing problems on hi-dpi screens.

wxGtk2 3.1.0 (shared library, not from Externals), KDE, Arch Linux

--

The dialog which is affected most of this is the GamePad config. Other dialogs become smaller too. In real-life I can't read anything, because everything is too small.

**master**
![screenshot_20160801_062929](https://cloud.githubusercontent.com/assets/13280758/17283558/230450e6-57b2-11e6-8f5b-7e156d7f067e.png)

**with hi-dpi patches**
![screenshot_20160801_063036](https://cloud.githubusercontent.com/assets/13280758/17283562/2eff4ac2-57b2-11e6-94bd-c90e64b49aeb.png)

EmptyChaos commented:
@GhettoGirl It looks like you DPI settings are messed up. Try opening a terminal and run:
```
xrandr --dpi 96x96
./dolphin-emu
```

GhettoGirl commented:
@EmptyChaos xrandr didn't changed anything so I guess my DPI settings are correct, otherwise even KDE whould look strange then. The dialogs are still very small.

Are you calculating the scaling factors correctly?

EmptyChaos commented:
Scaling factors are from WX (`FromDIP`). They're computed as `size * screen_dpi / 96`.

I was able to reproduce your screenshot on my test system by doing `xrandr --dpi 50x50` so I'm pretty sure there is something wrong with your setup (X11 tries to autodetect DPI and does a really crappy job sometimes [monitors can lie about their physical dimensions], it can pick values that are less than 96x96). You may want to try editing you xorg.conf and set the monitor DPI manually.

`xrandr` might not work if you're using the nvidia binary drivers.

GhettoGirl commented:
Hmm, very strange. I override the DPI setting in my xorg.conf now but nothing has changed. I'm pretty sure that my physical resolution is 96x96 dpi according to the manual of the monitor, so X11 detected this correctly beforehand without this override.

I wrote a quick test function in Qt and it returned 96. I should test this on wxWidgets too, maybe it is an upstream bug in wxWidgets?

--

Btw, I'm using Bumblebee. The same results on Intel and NVIDIA. But thats most likely not the issue.

EmptyChaos commented:
@GhettoGirl It's possible that Qt is just smarter about this than wx, it wouldn't be surprising.
```cpp
#include <stdio.h>
#include <wx/app.h>
#include <wx/dcscreen.h>

class TestApp : public wxApp
{
public:
	bool OnInit() override
	{
		if (!wxApp::OnInit())
			return false;
		
		wxSize ppi(wxScreenDC().GetPPI());
		printf("PPI: %dx%d\n", ppi.GetWidth(), ppi.GetHeight());
		return false;
	}
};
IMPLEMENT_APP(TestApp)
```

GhettoGirl commented:
The output is ```PPI: 96x96```, so wx detects this correctly.

EmptyChaos commented:
I'm now really confused. You can [see the implementation of FromDIP here](https://github.com/dolphin-emu/dolphin/blob/master/Externals/wxWidgets3/src/common/wincmn.cpp#L2860) and it hasn't changed in the latest wxWidgets so it should work. The images could break due to ContentScaleFactor but the only thing that affects layout is FromDIP. You could try sticking the PPI test in Dolphin and see if it gives a different result than the test app but it really shouldn't.

GhettoGirl commented:
The computed ContentScaleFactor is ``1.000000``. ```FromDIP``` returns the input.

EmptyChaos commented:
In that case it really should be working, the only major difference between this and master is that I ran all the pixel sizes through `FromDIP`. If `FromDIP` works (does nothing) then the result should be identical give or take a few incidental padding width changes. The DPI would have to change while Dolphin is running. What other UIs are affected as badly?

For InputConfigDiag, you could try setting [InputConfigDiag.cpp#831](https://github.com/EmptyChaos/dolphin/blob/wx-hidpi/Source/Core/DolphinWX/InputConfigDiag.cpp#L831) to:
```cpp
wxFont small_font(7, wxFONTFAMILY_DEFAULT, wxFONTSTYLE_NORMAL, wxFONTWEIGHT_NORMAL);
```
and [InputConfigDiagBitmaps.cpp#507](https://github.com/EmptyChaos/dolphin/blob/wx-hidpi/Source/Core/DolphinWX/InputConfigDiagBitmaps.cpp#L507):
```cpp
wxFont small_font(6, wxFONTFAMILY_DEFAULT, wxFONTSTYLE_NORMAL, wxFONTWEIGHT_BOLD);
```

GhettoGirl commented:
With that change the font size is the same as in **master** (in a readable size)
![screenshot_20160801_085041](https://cloud.githubusercontent.com/assets/13280758/17285768/3231e052-57c5-11e6-97f5-4a2756eb02dc.png)

Some buttons have a strange placement though :confused:

GhettoGirl commented:
In the Hotkey Settings Dialog the button placement is messed up too. The font size is identical (does this change affect the font size in this dialog too?)

EmptyChaos commented:
@GhettoGirl Okay, thanks for going through all this testing. The problem is with a change I made to the code, I changed the font sizes from point to pixels-plus-FromDIP and it seems your font engine decided that 5px tall is 'close enough' to 7px, which it really is not.

I'm not sure what I'll do about this, probably just rewrite all the affected code to use dialog units and point sizes since pixels and fonts go together like oil and water.

Yeah, the buttons are messed up because I somehow missed setting` wxALIGN_CENTER_VERTICAL` on all of them and only got most of them. It's trivial to fix. [Hotkeys and controllers use the same dialog, it just builds itself differently depending on the purpose it's serving]

GhettoGirl commented:
@EmptyChaos Glad I could help :smile: Otherwise I didn't noticed any other problems at the moment (there are really a lot of dialogs). If I find something new I will let you know.

EmptyChaos commented:
I believe I've resolved the problems with `InputConfigDialog`. I had been trying to avoid making changes to InputConfigDiagBitmaps up to this point but ultimately had to change it to use `wxGraphicsContext` instead of `wxDC`.

GhettoGirl commented:
The buttons are now probably aligned. Everything looks good now on my low-dpi screen :+1:

delroth commented:
Just tried the debugger with this PR and got a few "iCCP: known incorrect sRGB profile" errors pop up. Not sure if it's from this PR or if it's baseline.

EmptyChaos commented:
@delroth I have unreliably encountered that error on Linux myself but I haven't been unable to reproduce it consistently to debug it, I don't think it's specifically related to this PR. A quick Google indicates that it's a libpng error caused by PNG files with bad/corrupt iCCP (Color Profile) chunks. I'll see if I can analyze the PNGs.

EmptyChaos commented:
It's the debugger tool icons, they have iCCP chunks. (`Sys/Resources/toolbar_*`)

GitToTheHub commented:
Does it give any updates? I use a projector and generally changed the dpi under windows to 125%. The changes are looking very good. Would be great if this would be on master.

MaJoRoesch commented:
@EmptyChaos Would this benefit from new resources for different ratios?

EmptyChaos commented:
@MaJoRoesch Windows uses 125%, 150% and 200% scale ratios (OS X only uses 100%, 200% and 300%). To get the highest image quality it would be best to add `@1.25x` and `@1.5x` scale PNGs, but the existing `@2x` are probably good enough if we don't care about the runtime scaling down.

The only major problems are `isoproperties_disc/file/folder.png` and the `ratingX.png` files which don't have `@2x` versions.

MaJoRoesch commented:
Why don't they have @2x versions? I know I made and delivered them. :/

EmptyChaos commented:
@MaJoRoesch You may want to ping who you gave them to, they haven't been committed to the Git repository yet. (There shouldn't be any harm in committing them to master.)

MaJoRoesch commented:
@EmptyChaos They *should* have been committed, since I gave them! But let's see...

[•looks it up•](https://github.com/dolphin-emu/dolphin/commit/d86d5fae9f0bb5c0b588f58e1e45a5e025796903
)    -_-' Yay @Helios747...

@EmptyChaos I'll get you the @2x icons for those shortly!

EmptyChaos commented:
150%:
![mainsample150 opti](https://cloud.githubusercontent.com/assets/17393426/17645999/7ef81e58-61fa-11e6-90b1-68dd25d6fe12.png)

125%:
![mainsample125 opti](https://cloud.githubusercontent.com/assets/17393426/17646002/862b84da-61fa-11e6-98c9-6d1b924742d6.png)

Helios747 commented:
Yeah I didn't include them then because we weren't going to use them yet and had no idea when hidpi was going to be fixed @MaJoRoesch

RisingFog commented:
What's the status of this PR?

JosJuice commented:
I'm guessing the status is "nobody feels like reviewing it because it changes thousands of lines all across DolphinWX"

JosJuice commented on Source/Core/DolphinWX/WxUtils.cpp:
```diff
@@ -81,6 +81,39 @@ void AddToolbarButton(wxToolBar* toolbar, int toolID, const wxString& label, con
+  if (File::Exists(big_logo_path))
+    big_logo.LoadFile(StrToWxStr(big_logo_path), wxBITMAP_TYPE_PNG);
+
+  // Windows wants 16x16 (Window icon), 32x32/48x48 (Taskbar (big)) and 256x256 (Utility).
```
Is there a reason why Dolphin.ico can't be used for this? It already contains all those sizes.

In case we can't use that, we might want to consider including PNGs at the wanted sizes instead of scaling other PNGs.

shuffle2 commented:
![0f111fb8-541c-11e6-9747-b1832f5e559f](https://cloud.githubusercontent.com/assets/113063/19028579/996a50f2-88f2-11e6-9550-6da3ef5e8112.png)
Here, the second is the dialog after this commit...right? Doesn't the asymmetry of things like the C-Stick drawing seem gross?  
I use high-res monitors at 100% resolution so I don't see scaling issues...if other people think this commit makes the scaled case look better, then it seems fine to me - after rebase :)

EmptyChaos commented:
The screenshot at the top is old, that was just a scaling matrix applied to the existing drawing code. The current implementation is less blunt-force.
![icd150 opti](https://cloud.githubusercontent.com/assets/17393426/19030150/9476f1a4-8997-11e6-8732-61a7781640b7.png)

endrift commented:
I can test this at a handful of magnifications on Win10 and also on macOS tomorrow, if desired.

EmptyChaos commented:
@endrift If you could try this on OS X I'd appreciate it. I never got any feedback on how well it worked on OS X or not.

Could you also check if the dock-exit crash came back as well?

endrift commented:
I had figured @MaJoRoesch had tested it on OS X by now, but I was planning on doing it myself anyway. If you want testing on Linux/KDE too, I can manage that.

lioncash commented:
What do the two commits before the last commit have to do with HiDPI support? If the answer is essentially "Nothing in particular", please separate them out into their own PRs after this is merged or something. It makes it way easier to review, and keeps unrelated changes separate.

EmptyChaos commented:
@endrift I believe it should work fine on Linux as I've tested it; Dolphin only supports wxGTK, not wxQt, so it shouldn't behave differently on KDE.

@lioncash It's been a while so I don't really remember, I think I was having problems with deadlocks in GTK at one point. It doesn't matter, I've dropped the commits.

MaJoRoesch commented:
@endrift I haven't tested it since I don't have a retina macbook. I could, by using my 1440p monitor and the hack I have enabled, but that's a bit of work.

shuffle2 commented:
In windows it is just a setting. Isn't it the same on OS X?  

![](http://www.dotnetmafia.com/blogs/dotnettipoftheday/W10B10074DisplayScalingSurfacePro3_669BCFD0.png)

EmptyChaos commented:
@shuffle2 That should be the DPI setting on Windows 10, yes.

shuffle2 commented:
@EmptyChaos yea i'm just pointing out that @MaJoRoesch probably doesn't need a special screen to test.

EmptyChaos commented:
I think OS X has a "more space"/"bigger text" setting somewhere in the control panel that functions the same way.

MaJoRoesch commented:
@shuffle2 It is not a setting on MacOS. It uses a whitelist, and you have to use what is essentially a registry hack to allow scaling on 3rd party monitors. Also, it requires 1440p or above to scale.

shuffle2 commented:
welp

aldelaro5 commented:
After applying this PR, I found 2 new issues about the debugger UI.

The first one is the memory view, it doesn't load the debugger font, evrything else loads it sucessfully.

The second issue is the spacing of the different parts of instructions.  More precisely, there's 2 places where the spacing is not what it shoudl be.

When using any non fixed width font (tested with noto sans, serif and a few others such as arial, as long as it's not fixed width), the address and the opcode name don't have any spacing between each other.  On monospace and consolas however, this doesn't seem to be an issue.

For comparision, here's with arial:
![screenshot from 2016-10-04 02-07-09](https://cloud.githubusercontent.com/assets/8932978/19064331/accd53c2-89da-11e6-8467-8c9da7f5c383.png)

and with consolas:
![screenshot from 2016-10-04 02-07-33](https://cloud.githubusercontent.com/assets/8932978/19064315/8fd84a56-89da-11e6-8a37-82c12029ee3c.png)

The second place is between the symbol name and the branches lines, they actually overlap each other no matter the font:
![screenshot from 2016-10-04 02-07-54](https://cloud.githubusercontent.com/assets/8932978/19064299/79c41fb0-89da-11e6-8c6a-4774003f89a0.png)

EmptyChaos commented:
@aldelaro5 #4290

shuffle2 commented on Source/Core/DolphinWX/LogConfigWindow.cpp:
```diff
@@ -25,18 +25,11 @@ LogConfigWindow::LogConfigWindow(wxWindow* parent, wxWindowID id)
 
-void LogConfigWindow::OnClose(wxCloseEvent& event)
-{
-  SaveSettings();
```
This was removed without adding the call back somewhere, so Log configuration settings are currently not being persisted into the .ini file.

shuffle2 replied:
Good point!

shuffle2 replied:
Just brought the old code back in https://github.com/dolphin-emu/dolphin/pull/4292
